lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -684,6 +805,88 @@ def _set_unit_ip(self, event):
                 self._charm.model.get_binding(relation).network.bind_address
             )
 
+    def _label_alert_topology(self, rule):
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing Prometheus alert rule with Juju
+            topology labels.
+        """
+        metadata = self._scrape_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule):
+        """Insert juju topology filters into a Prometheus alert rule.
+
+        Args:
+            rule: a dictionary representing a Prometheus alert rule.
+
+        Returns:
+            a dictionary representing a Prometheus alert rule that filters based
+            on juju topology.
+        """
+        metadata = self._scrape_metadata
+        topology = (
+            'juju_model="{}", juju_model_uuid="{}", juju_application="{}"'.format(
+                metadata["model"], metadata["model_uuid"], metadata["application"]
+            )
+        )
+
+        if expr := rule.get("expr", None):
+            expr = expr.replace("%%juju_topology%%", topology)
+            rule["expr"] = expr
+        else:
+            logger.error("Invalid alert expression in %s", rule.get("alert"))
+
+        return rule
+
+    @property
+    def _labeled_alert_groups(self):
+        """Load alert rules from rule files.
+
+        All rules from files for a consumer charm are loaded into a single
+        group. The generated name of this group includes Juju topology
+        prefixes.
+
+        Returns:
+            a list of Prometheus alert rule groups.
+        """
+        alerts = []
+        for path in Path(ALERT_RULES_PATH).glob("*.rule"):
+            if not path.is_file():
+                continue
+
+            with path.open() as rule_file:
+                # Load a list of rules from file then add labels and filters
+                try:
+                    rules = yaml.safe_load(rule_file)
+                    rule = rules[0]  # each file is list of one rule
+                    rule = self._label_alert_topology(rule)
+                    rule = self._label_alert_expression(rule)
+                    alerts.append(rule)
+                except Exception:
+                    logger.error("Failed to read alert rules from %s", path.name)
```

Comment:
This is typically a linting error on PyCharm: `except Exception` may be used only if followed by `raise`; otherwise should be specific on the exception types.
For example you can specify yaml.scanner.ScannerError, IndexError, etc.

Reply:
In general it would be a good idea to trap specific exceptions. In this case it is largely pointless since all errors are handled in the same way.

Reply:
Also this is why I think there is no substitute for human over sight such as code review. Automated tools are just that and hence can be pedantic.